dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java

```diff
@@ -0,0 +1,164 @@
+    SingerMessage recordMessage2 =
+        MessageUtils.createRecordMessage(TABLE_NAME, COLUMN_NAME, "yellow");
+
+    final List<SingerMessage> expected = Lists.newArrayList(recordMessage1, recordMessage2);
+
+    StandardTapConfig tapConfig =
+        WorkerUtils.syncToTapConfig(TestConfigHelpers.createSyncConfig().getValue());
+    StandardDiscoverSchemaInput discoverSchemaInput = new StandardDiscoverSchemaInput();
+    discoverSchemaInput.setConnectionConfiguration(
+        tapConfig.getSourceConnectionImplementation().getConfiguration());
+    ProcessBuilderFactory pbf = mock(ProcessBuilderFactory.class);
+    ProcessBuilder processBuilder = mock(ProcessBuilder.class);
+    Process process = mock(Process.class);
+    InputStream inputStream = mock(InputStream.class);
+    SingerDiscoverSchemaWorker discoverSchemaWorker = mock(SingerDiscoverSchemaWorker.class);
+
+    final SingerColumn singerColumn = new SingerColumn();
+    singerColumn.setType(Lists.newArrayList(SingerType.NULL, SingerType.STRING));
+
+    final SingerColumnMap singerColumnMap = new SingerColumnMap();
+    singerColumnMap.setAdditionalProperty(COLUMN_NAME, singerColumn);
+
+    final SingerTableSchema singerSchema = new SingerTableSchema();
+    singerSchema.setType("object");
+    singerSchema.setProperties(singerColumnMap);
+
+    SingerStream singerStream = new SingerStream();
+    singerStream.setStream(TABLE_NAME);
+    singerStream.setTapStreamId(TABLE_NAME);
+    singerStream.setTableName(TABLE_NAME);
+    singerStream.setSchema(singerSchema);
+
+    final SingerCatalog singerCatalog = new SingerCatalog();
+    singerCatalog.setStreams(Collections.singletonList(singerStream));
+
+    final OutputAndStatus<SingerCatalog> discoverSchemaOutput =
+        new OutputAndStatus<>(JobStatus.SUCCESSFUL, singerCatalog);
+
+    when(discoverSchemaWorker.runInternal(discoverSchemaInput, discoverSchemaJobRoot))
+        .thenReturn(discoverSchemaOutput);
+
+    when(pbf.create(
+            jobRoot,
+            IMAGE_NAME,
+            "--config",
+            CONFIG_JSON_FILENAME,
+            "--properties",
+            CATALOG_JSON_FILENAME,
+            "--state",
+            STATE_JSON_FILENAME))
+        .thenReturn(processBuilder);
+    when(processBuilder.redirectError(errorLogPath.toFile())).thenReturn(processBuilder);
+    when(processBuilder.start()).thenReturn(process);
+    when(process.getInputStream()).thenReturn(inputStream);
+
+    when(streamFactory.create(any())).thenReturn(expected.stream());
+
+    final SingerTapFactory tapFactory =
+        new SingerTapFactory(IMAGE_NAME, pbf, streamFactory, discoverSchemaWorker);
+    final Stream<SingerMessage> actual = tapFactory.create(tapConfig, jobRoot);
+
+    assertTrue(Files.exists(jobRoot));
+    assertTrue(Files.exists(jobRoot.resolve("tap_config.json")));
+    assertTrue(Files.exists(jobRoot.resolve("catalog.json")));
+    assertTrue(Files.exists(jobRoot.resolve("input_state.json")));
```

Comment:
i should check that the actual contents of these files are what i expect them to be.